util/types/src/core/extras.rs

```diff
@@ -105,80 +143,91 @@ impl EpochExt {
         self.difficulty = difficulty;
     }
 
-    pub fn difficulty(&self) -> &U256 {
-        &self.difficulty
+    //
+    // Normal Methods
+    //
+
+    pub fn new_builder() -> EpochExtBuilder {
+        EpochExtBuilder(EpochExt::default())
     }
 
-    pub fn remainder_reward(&self) -> &Capacity {
-        &self.remainder_reward
+    pub fn into_builder(self) -> EpochExtBuilder {
+        EpochExtBuilder(self)
     }
 
-    pub fn last_block_hash_in_previous_epoch(&self) -> packed::Byte32 {
-        self.last_block_hash_in_previous_epoch.clone()
+    pub fn is_genesis(&self) -> bool {
+        0 == self.number
     }
 
-    pub fn previous_epoch_hash_rate(&self) -> &U256 {
-        &self.previous_epoch_hash_rate
+    pub fn block_reward(&self, number: BlockNumber) -> Result<Capacity, FailureError> {
+        if number >= self.start_number()
+            && number < self.start_number() + self.remainder_reward.as_u64()
+        {
+            self.base_block_reward
+                .safe_add(Capacity::one())
+                .map_err(Into::into)
+        } else {
+            Ok(self.base_block_reward)
+        }
     }
+}
+
+impl EpochExtBuilder {
```

Comment:
I think the builder is redundant

Reply:
I didn't get your point.

IMHO, I think we should remove all setters from `EpochExt`.

Reply:
See this:
https://github.com/nervosnetwork/ckb/blob/87c6912deb4427f8ac32cfe54d5b1bea3ca44da6/spec/src/consensus.rs#L211-L218

If we calling `set_genesis_block` before calling `set_genesis_epoch_ext`, a bug was created.

https://github.com/nervosnetwork/ckb/blob/ecd88dcffe7c18c7112d2d4f1cce71560ba2618d/spec/src/lib.rs#L189-L190

Two independent methods have to be used in a specific order.

I think we can remove all `setters` from `EpochExt` to avoid that. (We can do it after `ConsensusBuilder`, the `ConsensusBuilder` is on the task list.)
If u want to update a `EpochExt`, u can do `ext.into_builder().set_xxx().build()`.